Retry requests if storage (auth + credentials) is not ready

marcochavezf opened the pull request:
cc @marcaaron @tgolen 

<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

There's an [issue](https://github.com/Expensify/App/issues/5619) on Android where the user is logged out randomly. We're not 100% sure about the root cause, but one possibility is the existence of a race condition between the async storage and the networks requests; where the auth token and credentials are not read yet from persistent storage (Onyx) while a network request is made.

I was able to reproduce this scenario by adding a `setTimeout` in the `Onyx.connect` callback (more details in this [comment](https://github.com/Expensify/App/issues/5619#issuecomment-941751331)):

https://user-images.githubusercontent.com/6829422/137043249-89fe6943-ce38-452e-8704-a509d5ebdcef.mov

There's a long discussion in the [issue](https://github.com/Expensify/App/issues/5619), where different approaches are discussed (both big and small changes). But we agreed on tackling the issue atm without changing the architecture of the Network Queue. This solution is comprised of the following points:

- Add a unit test (`StoreNetworkTest.js`) that simulates the scenario where auth token + credentials are not ready.
- Add the flag `isReady` and its setter to `Network.js` to process requests only if auth token + credentials are already read from storage, otherwise re-queue the requests to retry later (as long as the request don't have the property `doNotRetry` = true). Also, log a message to the server when this scenario happens.
- Stop logging out in `enhanceParameters` (which is `addDefaultValuesToParameters` in `API.js`), since we're already handling that auth token + credentials are read first and the http request should logout the user if the auth token is inexistent or expired (server should return a 407 `jsonCode` if we are online). Also, the `API.js` has changed a lot since the logout (`redirectToSignIn()`) was [added](https://github.com/Expensify/App/pull/583/files) and probably is no longer needed because that's already handled in `handleExpiredAuthToken`.

Also a few extra comments:
- In order to use the `Log` in `Network` I disabled the eslint circular dependency warning (hopefully we can fix this later with a better dependency or another design).
- Fixed a few minor jsdocs types.
- Added explicit Jest globals to get rid of errors in vscode and have better IntelliSense.
- I updated some unit tests to verify the retry is not happening (which is an ideal scenario where auth token + credentials are already read when a network request happens).
- Also added `Onyx.init` and `Onyx.clear` in some tests, because for some reason the `Onyx.connect` callback was not triggered and those tests were failing.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/5619

### Tests / QA Steps
<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
In Chrome devtools go to Application -> Local Storage and: 
- Mess up the `authToken` (session) and make sure you are reauthenticated.
- Mess up the `credentials` and `authToken` and make sure you are logged out.

<img width="800" alt="Screen Shot 2021-11-08 at 11 22 45" src="https://user-images.githubusercontent.com/6829422/140779358-27a2d72f-5da3-4ca9-80b1-b7aa49856c55.png">

<!--- 
### QA Steps
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
 N/A

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
 N/A

marcochavezf committed:
retry a request if auth+crendentials are not ready

marcochavezf committed:
disable circular dep between Log-Api-Network

marcochavezf committed:
- add Onyx initialization
- add explicit globals to improve intellisense in vscode

marcochavezf committed:
add unit test to check if auth+cred are not ready

marcochavezf committed:
add documentation to isReady
move isReady to canMakeRequest

marcochavezf committed:
stop logging the user out in enhanceParameters
(let the http request logout if the auth token is expired or inexistent)

marcaaron commented on tests/actions/ReportTest.js:
```diff
@@ -30,8 +34,15 @@ describe('actions/Report', () => {
+
+        Onyx.init({
+            keys: ONYXKEYS,
+            registerStorageEventListener: () => { },
```
```suggestion
            registerStorageEventListener: () => {},
```

marcaaron commented on tests/actions/SessionTest.js:
```diff
@@ -16,6 +17,13 @@ jest.mock('../../src/libs/Notification/PushNotification', () => ({
 
+Onyx.init({
+    keys: ONYXKEYS,
+    registerStorageEventListener: () => { },
```
```suggestion
    registerStorageEventListener: () => {},
```

marcaaron commented on src/libs/Network.js:
```diff
@@ -95,15 +98,33 @@ Onyx.connect({
     callback: val => email = val ? val.email : null,
 });
 
+/**
+ *
+ * @param {Boolean} val
+ */
```
```suggestion
/**
 * @param {Boolean} val
 */
```

marcaaron commented on tests/unit/StoreNetworkTest.js:
```diff
@@ -0,0 +1,47 @@
+
+Onyx.init({
+    keys: ONYXKEYS,
+    registerStorageEventListener: () => { },
```
```suggestion
    registerStorageEventListener: () => {},
```

marcochavezf committed:
Update tests/unit/StoreNetworkTest.js

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

marcochavezf committed:
Update src/libs/Network.js

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

marcochavezf committed:
Update tests/actions/SessionTest.js

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

marcochavezf committed:
Update tests/actions/ReportTest.js

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

marcochavezf commented on src/libs/API.js:
```diff
@@ -1,26 +1,43 @@
+ * Checks if both authToken and credentials were already read from persistent storage (Onyx)
+ * @returns {Boolean}
+ */
+function isAuthStoreReady() {
```
Good call, I will change it to that function 👍🏽

marcochavezf commented on src/libs/API.js:
```diff
@@ -58,8 +75,6 @@ function addDefaultValuesToParameters(command, parameters) {
         // If we end up here with no authToken it means we are trying to make an API request before we are signed in.
         // In this case, we should cancel the current request by pausing the queue and clearing the remaining requests.
         if (!authToken) {
-            redirectToSignIn();
```
Good question, it makes sense to remove that code too because that case is already handled with the `isReady` flag and we shouldn't block the request (pausing or clearing it) here if the logout is handled by the http request (407 `jsonCode` server response) when the `authToken` is expired or non-existent.

Also, I think we also should remove this code, no?

https://github.com/Expensify/App/blob/8ccf3b1a5ae1b0286b21e0bd95b4af80615bc936/src/libs/Network.js#L203-L208

I will remove all that code and test.

marcaaron replied:
Yeah I _think_ we can. My reasoning is something like:

1. We are now guaranteeing that both `authToken` and `credentials` are read from storage
2. If we are online and have no `authToken` but do have `credentials` then we should `reauthenticate`
3. If we are online and have no `authToken` or `credentials` then we should get signed out by `reauthenticate`

marcaaron replied:
The only thing that should "pause" the network request queue is the process of authenticating. But we should test this assumption to be sure it's correct.

marcochavezf replied:
I tested without that code and everything seems to be working fine (I tested deleting and modifying the auth token and credentials from the browser local storage).

marcaaron replied:
>deleting and modifying the auth token

This should cause a reauthentication

>and credentials from the browser local storage

If we also do this then we should get logged out.

To clarify, is that what you mean by "working fine" or something else?

marcochavezf replied:
Ah yeah, when I was deleting or modifying the auth token the reauthentication was triggered and also when I was wiping out the credentials + auth token the user was logged out

marcochavezf committed:
add suggestions to API.js and StoreNetworkTest.js

marcochavezf commented on tests/unit/StoreNetworkTest.js:
```diff
@@ -0,0 +1,47 @@
+beforeEach(() => Onyx.clear().then(waitForPromisesToResolve));
+
+// This unit test was moved to its own file because modifies Network.setIsReady which
+// could mess up other unit tests, and also beforeEach can't be used for this scenario.
```
Yeah, I will check if resetting the state of `isReady` does the trick. Also, I think using manual mock or mocking `AsyncStorage.getItem()` could fix the issue of having this test in `NetworkTest` without messing with the other unit tests.

marcochavezf commented on src/libs/Network.js:
```diff
@@ -95,15 +98,33 @@ Onyx.connect({
+    // forceNetworkRequest and isQueuePaused because we wait until credentials + session objects have
+    // been read from persistent storage).
+    if (!isReady) {
+        LogUtil.hmmm('Trying to make a request when Network is not ready', {command: request.command, type: request.type});
```
Hehe yeah, I think it can help us to know which command is being called before auth + credentials are read from Onyx. Also in an ideal scenario that log shouldn't be called in unit tests (I added a couple of `expect` statements to check that message is not logged if auth token + credentials are already read).

marcaaron replied:
So, what I was getting at here is that...

1. We should use a constant whenever referencing something like an error message in multiple places. That way if the content changes later the reference won't.
2. Logs are not great things to test against because people can decide to remove them and the tests that are expecting them "not to be called" are going to keep passing anyway.

marcochavezf replied:
Ah right, for the point 2, I removed the tests that check if the Log was not called with that message.

marcaaron replied:
NAB, `request.type` can probably be removed it's almost always going to be a `post` and not really relevant

marcochavezf commented on src/libs/Network.js:
```diff
@@ -95,15 +98,33 @@ Onyx.connect({
 function canMakeRequest(request) {
+    // We'll not make the request now if the network is not ready (it should have higher priority than
+    // forceNetworkRequest and isQueuePaused because we wait until credentials + session objects have
+    // been read from persistent storage).
```
Yeah, if we look at the code now it's easy to understand the priority because the context is still fresh in our minds, but maybe 6 months later could not be obvious for someone else (or even ourselves 😀). Also, I think that comment could help someone else (who wants to modify the `canMakeRequest`) to get up to speed quickly with the context. We can modify or remove the comment if `canMakeRequest` is changed again later :)

marcaaron replied:
>maybe 6 months later could not be obvious for someone else

On the other hand, all the context is here in GitHub. But if we are worried about this behavior changing it might be a good thing to write a test for.

marcaaron replied:
>We can modify or remove the comment if canMakeRequest is changed again later :)

I think the case I'm worried about is that we have to remember to update the comment (and it's very possible we will).

marcochavezf replied:
I removed the comment and updated the unit test

marcochavezf committed:
remove code to pause and clear queue in
enhanceParameters and its related canMakeRequest call

marcochavezf committed:
add delay to AsyncStorage.getItem()
move StoreNetworkTest to NetworkTest

marcochavezf committed:
remove comment in canMakeRequest
and update unit test

marcochavezf committed:
check if the network request is retried in test

marcochavezf committed:
restore mock async-storage to its original verison

marcochavezf committed:
add Onyx mock to add delay in connect callbacks

marcochavezf committed:
fix jsdoc param for GetAccountStatus

marcochavezf committed:
fix test to retry request if Network is not ready

marcochavezf committed:
Revert "add missing optional param in jsdoc"

This reverts commit ec2ee0874b80a4c38a0535b07493c07f45f0b774.

marcochavezf committed:
add missing optional param in jsdoc

nickmurray47 reviewed (commented):
The changes lgtm afaict - looks like there's one linting issue and also wondering if we can add manual tests for regressions to make sure we aren't breaking anything.

marcochavezf committed:
Revert "add waitForPromisesToResolve in test"

This reverts commit 62ef02048c18e8e00d0618cd1ed25050aeeef700.

marcochavezf committed:
add waitForPromisesToResolve in test

marcaaron commented on __mocks__/react-native-onyx.js:
```diff
@@ -0,0 +1,24 @@
+    }),
+    addDelayToConnectCallback,
+};
+export {withOnyx};
```
Hmm I liked the mock we had before. Why does it not work?

marcochavezf replied:
Yeah, I was trying to make it work with AsyncStorage mock but I found a rabbit hole when it's used with Onyx. The main problem is that `API.js` already calls `Onyx.connect` before the unit tests are set up, which makes it a little difficult to have control of those `Onyx.connect` calls. For example in some cases `matchingKeys` were empty (I think because of Onyx.clear in beforeEach) and in other cases  the`Onyx.connect` callbacks registered at the top of `API.js` were already cached and not being resolved by the `AsyncStorage.getItem` function

https://github.com/Expensify/react-native-onyx/blob/224e2ba03f2365726b8c1d0785e750c54a85332b/lib/Onyx.js#L42-L45

Maybe an easier solution (not for now) could be an `init()` function in `API.js`, which register the `Onyx.connect` callbacks, and we can call it after preparing / setting up the unit tests.

marcaaron replied:
Ah too bad. I guess it's OK? I found this a little confusing because we are not adding the throttle to the thing that we suspect is going slow, but to `Onyx.connect()` which can't really be slow.

marcochavezf replied:
Yeah, I think mocking only `Onyx.connect()` is ok atm to only test the network retry when storage is not ready. Hopefully, we can revise this test later and mock `AsyncStorage. getItem` after agreeing something in the `Onyx.connect` [discussion](https://github.com/Expensify/App/issues/6151).

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -541,7 +541,7 @@ function getOptions(reports, personalDetails, draftComments, activeReportID, {
  * @param {Object} personalDetails
  * @param {String} searchValue
- * @param {Array<String>} betas
+ * @param {Array<String>=} betas
```
Whats the `=` doing here? Is this intentional?

marcaaron replied:
Interesting.. I don't get that error and use vscode - perhaps you have some global rules set up.

marcaaron replied:
Forgot this one

marcochavezf replied:
I think was this [extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) which showed me that warning in vscode 👍🏽

marcaaron replied:
Let's remove it?

marcochavezf replied:
Ok sure, I will revert it

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+    Onyx.addDelayToConnectCallback(3000);
+
+    // Reset isReady for this unit test, because Onyx.clear() in beforeEach resolves the
+    // Onyx.connect callbacks with null values
```
I think it would be good to adopt the `GIVEN` `WHEN` `THEN` style of comments here. There's some information on that here:

https://github.com/Expensify/App/blob/main/tests/README.md#documenting-tests

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+
+    // Set values to trigger Onyx.connect callbacks
+    Onyx.merge(ONYXKEYS.CREDENTIALS, {});
+    Onyx.merge(ONYXKEYS.SESSION, {});
```
I'm confused about why we are merging empty objects here and what the initial state of these values should be.

marcaaron replied:
e.g. I think the bug we are observing has the values for `authToken` and `credentials` already existing, but just taking a long time to be read from storage. So I'm unsure what we are trying to simulate by merging in the empty object..

marcochavezf replied:
I'm merging an empty object just to force the `Onyx.connect` callbacks registered in `API.js` (which kind of workaround the problem of the previous [comment](https://github.com/Expensify/App/pull/6183#discussion_r743222649)). And it's an empty object because in this unit test we only want to verify that a network request is retried when the `authToken` (which can be non-existent, valid, or expired) and `credentials` haven't been read from the storage.

Also, I think the initial values are empty because of the `Onyx.clear` call in `beforeEach`, that's why I thought made sense to merge only an empty object in both.

marcaaron replied:
I think I'm following... The comment does explain that we want trigger the callback, but I was unsure why we chose to merge an empty object into these keys to do that. Thinking about it some more I guess the answer is... it's just _something_ that won't affect the value and will trigger an update. And we are having difficulty creating a situation where `Onyx.connect()` has not yet run.

marcochavezf replied:
I added a comment to explain this better and why we merge an empty object 👍🏽

marcaaron replied:
Great thanks!

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+
+    // Given a test user login and account ID
+    const TEST_USER_LOGIN = 'test@testguy.com';
+    fetchAccountDetails(TEST_USER_LOGIN);
```
Why do we need to `fetchAccountDetails()`? Are we just making some arbitrary request that requires an `authToken`?

marcochavezf replied:
Yeap, exactly, it's just an arbitrary request which can be retried if the auth token + credentials are not read from storage yet.

marcaaron replied:
That would be a good thing to note maybe in a comment.

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+        expect(spyHttpUtilsXhr).not.toHaveBeenCalled();
+
+        // Resolve Onyx.connect callbacks
+        jest.runOnlyPendingTimers();
```
Would it be clearer to `jest.advanceTimersByTime(ONYX_DELAY_MS)`?

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
 });
+
+test('retry network request if auth and credentials are not read from Onyx yet', () => {
+    Onyx.addDelayToConnectCallback(3000);
```
```suggestion
    const ONYX_DELAY_MS = 3000;
    Onyx.addDelayToConnectCallback(ONYX_DELAY_MS);
```

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,45 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+        expect(spyHttpUtilsXhr).not.toHaveBeenCalled();
+
+        // Run processNetworkRequestQueue in the setInterval of Network.js
+        jest.runOnlyPendingTimers();
```
Same question here - since if we want to advance to get the Network loop to tick then it's 1 second (and we could also set up a constant for that).

marcochavezf replied:
Sure, I will change it to `jest.advanceTimersByTime(NETWORK_INTERVAL_MS);` where `NETWORK_INTERVAL_MS = 1000`

marcaaron commented on src/libs/API.js:
```diff
@@ -55,18 +71,6 @@ function addDefaultValuesToParameters(command, parameters) {
-            Network.clearRequestQueue();
-            Network.unpauseRequestQueue();
-            return;
-        }
```
@nickmurray47 comment makes me wonder if we should add a test for this. But maybe it is already covered.

marcaaron replied:
I guess the manual test would just be...

- Mess up the `authToken` and make sure you are reauthenticated
- Mess up the `credentials` and `authToken` and make sure you are logged out

nickmurray47 replied:
I like adding both of those!

marcaaron commented:
This is testing pretty good for me manually - doesn't seem like there are any major regressions. I did have one question though...

Since we are setting the Network to "ready" when the `session` and `credentials` are available should we set it back to "not ready" when we sign out? If the Network was still in the "ready" state and we tried to sign in I'm curious if we would run into any issues if e.g. their local variables were not yet set? 🤔

marcochavezf committed:
fix callback params in mock Onyx

marcochavezf commented:
> Since we are setting the Network to "ready" when the `session` and `credentials` are available should we set it back to "not ready" when we sign out? If the Network was still in the "ready" state and we tried to sign in I'm curious if we would run into any issues if e.g. their local variables were not yet set? 🤔

Hmm I don't think so, because the `Network.setIsReady(true)` should be only called when the app is booting up to avoid the race condition between storage and network. Once `session` and `credentials` (existent or non-existent) are read from storage the Network should be ready until the app is closed by the OS. I think we should only set the network to "not ready" if the storage fails for some reason while the app is still running on a handset (not sure if that's something the OS handles making the app crash, but in any case should be a rare scenario).

marcochavezf committed:
change runOnlyPendingTimers to advanceTimersByTime

marcaaron commented:
>Hmm I don't think so, because the Network.setIsReady(true) should be only called when the app is booting up to avoid the race condition between storage and network.

Yeah I was thinking that when you sign out the storage is cleared so the values for `session` and `credentials` look like this:

<img width="507" alt="2021-11-04_12-41-48" src="https://user-images.githubusercontent.com/32969087/140430271-59e7bf7f-1d3d-4920-8ace-5f99af18076e.png">

But since we "cache-first" there's no way for an `Onyx.set()` or `Onyx.merge()` to cause a "slow" `Onyx.connect()` on sign in.

https://github.com/Expensify/react-native-onyx/blob/224e2ba03f2365726b8c1d0785e750c54a85332b/lib/Onyx.js#L530-L537

marcaaron commented on tests/unit/NetworkTest.js:
```diff
@@ -209,3 +221,47 @@ test('consecutive API calls eventually succeed when authToken is expired', () =>
+
+test('retry network request if auth and credentials are not read from Onyx yet', () => {
+    const ONYX_DELAY_MS = 3000;
+    const NETWORK_INTERVAL_MS = 1000;
```
This can be changed by `Network` so I was thinking that we'd add this in `Network.js`

marcochavezf replied:
> > Hmm I don't think so, because the Network.setIsReady(true) should be only called when the app is booting up to avoid the race condition between storage and network.
> 
> Yeah I was thinking that when you sign out the storage is cleared so the values for `session` and `credentials` look like this:
> 
> <img alt="2021-11-04_12-41-48" width="507" src="https://user-images.githubusercontent.com/32969087/140430271-59e7bf7f-1d3d-4920-8ace-5f99af18076e.png">
> 
> But since we "cache-first" there's no way for an `Onyx.set()` or `Onyx.merge()` to cause a "slow" `Onyx.connect()` on sign in.
> 
> https://github.com/Expensify/react-native-onyx/blob/224e2ba03f2365726b8c1d0785e750c54a85332b/lib/Onyx.js#L530-L537

Ah yeah but even if we don't read the values from the cache after signing out, I don't think we should have a problem; the storage and the network are ready and the network loop should handle any bad state.

marcochavezf replied:
I added the `PROCESS_REQUEST_DELAY_MS` const to `Network.js` and it's imported in `NetworkTest.js`

marcochavezf committed:
add PROCESS_REQUEST_DELAY_MS to Network
change unit test comments to GIVE-WHEN-THEN format

marcochavezf commented:
> > Hmm I don't think so, because the Network.setIsReady(true) should be only called when the app is booting up to avoid the race condition between storage and network.
> 
> Yeah I was thinking that when you sign out the storage is cleared so the values for `session` and `credentials` look like this:
> 
> <img alt="2021-11-04_12-41-48" width="507" src="https://user-images.githubusercontent.com/32969087/140430271-59e7bf7f-1d3d-4920-8ace-5f99af18076e.png">
> 
> But since we "cache-first" there's no way for an `Onyx.set()` or `Onyx.merge()` to cause a "slow" `Onyx.connect()` on sign in.
> 
> https://github.com/Expensify/react-native-onyx/blob/224e2ba03f2365726b8c1d0785e750c54a85332b/lib/Onyx.js#L530-L537

Ah yeah but even if we don't read the values from the cache after signing out, I don't think we should have a problem; the storage and the network are ready and the network loop should handle any bad state.

marcaaron reviewed (dismissed):
LGTM 👍  

It would be nice to see if the test can be improved further, but maybe not crucial to do right now. I think what we've come up works OK for now.

marcaaron commented:
Maybe merging main will fix the e2e test or we can restart it.

marcochavezf committed:
update comment in test that retrires network req

marcochavezf commented:
> It would be nice to see if the test can be improved further, but maybe not crucial to do right now. I think what we've come up works OK for now.

Yeah, hopefully, we can improve later this test (which I think we'll require a few changes in `API.js` to have better control of the `Onyx.connect` callbacks). Also in the coverage report (running `npm run test -- --coverage`) I noticed we cover 75.58% of `Network.js` code in the unit tests:

<img width="800" alt="Screen Shot 2021-11-05 at 12 40 07" src="https://user-images.githubusercontent.com/6829422/140562427-1e8fbd48-2356-42cd-84c7-3b8d407fb974.png">

And given the importance of `Network.js` in the app, should we add more unit tests to cover if possible 100% in this file? 🤔

> Maybe merging main will fix the e2e test or we can restart it.

I added another commit with more comments in the unit test and now the e2e test passed 😄

marcaaron commented:
Left a follow up comment about the JS doc, but I think this is good to merge.

nickmurray47 commented:
Not sure if tests failing had anything to do with the recent change but restarted them to see if they'll pass again

nickmurray47 reviewed (dismissed):
Tests pass now! Feel free to merge and @marcochavezf do you mind adding these QA steps to the PR from [this comment](https://github.com/Expensify/App/pull/6183#discussion_r743139411):
- Mess up the `authToken` and make sure you are reauthenticated
- Mess up the `credentials` and `authToken` and make sure you are logged out

marcochavezf committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

Conflicts:
	src/libs/Network.js

marcochavezf committed:
remove duplicated lodashGet

marcochavezf committed:
disable prefer-onyx-connect-in-libs in Onyx mock

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1463912379) to staging by @marcaaron in version: 1.1.14-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

isagoico commented:
@marcaaron @marcochavezf Hello! I'm not sure if were testing this correctly. What I did was: 
1. Navigate to Application in dev tools 
2. In local storage search for session 
3. In the session info I changed the Authtoken and the email. I just added a bunch of random characters to mess it up 

After that, I was **not** logged out and it looked like I was logged in as the user I randomly wrote on the email
![image](https://user-images.githubusercontent.com/44479856/142003014-8f2f9314-d482-4782-b1e6-16b67719ce0a.png)

marcochavezf commented:
> @marcaaron @marcochavezf Hello! I'm not sure if were testing this correctly. What I did was:
> 
> 1. Navigate to Application in dev tools
> 2. In local storage search for session
> 3. In the session info I changed the Authtoken and the email. I just added a bunch of random characters to mess it up
> 
> After that, I was **not** logged out and it looked like I was logged in as the user I randomly wrote on the email ![image](https://user-images.githubusercontent.com/44479856/142003014-8f2f9314-d482-4782-b1e6-16b67719ce0a.png)

Ah from the app side I think this is fine because we're only verifying that `authToken` (from session object) and the `credentials` object (which is another object in the local storage) are valid or existent to keep the user logged in (which are retrieved from the server when the user enters their correct email and password). And the `email` from the session object is only used for the UI and not used to re-login the user.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1491852026) to production by @AndrewGable in version: 1.1.15-15 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
